File: c# Dotnet Core/STED/src/UseCasesTests/UserAccessCheckerTest.cs
using Domain;
using Moq;
using Tools;
using UseCases;

namespace UseCasesTests
{
    public class UserAccessCheckerTest
    {
        [Fact]
        public void Test_No_Name_No_Password_should_return_Argument_Exception()
        {
            IUser userToCheck = new User();
            Mock<IUserByNameLoader> loader = new Mock<IUserByNameLoader>();
            loader.Setup(l=>l.Load("")).Returns<IUser?>(null);

            Assert.Throws<ArgumentException>(() => new UserAccessChecker(loader.Object, new Tools.RandomString()).Check(userToCheck));
        }

        [Fact]
        public void Test_Unknown_Name_should_return_Error_Argument_Exception()
        { 
            IUser userToCheck = new User { Name="debra" };
            Mock<IUserByNameLoader> loader = new Mock<IUserByNameLoader>();
            loader.Setup(l => l.Load("debra")).Returns<IUser?>(null);

            Assert.Throws<ArgumentException>(() => new UserAccessChecker(loader.Object, new Tools.RandomString()).Check(userToCheck));
        }

        [Fact]
        public void Test_known_Name_With_Known_Password_should_return_Valid_Token()
        { 
            IUser userToCheck = new User { Name = "dexter", Password = "killer" };
            Mock<IUserByNameLoader> loader = new Mock<IUserByNameLoader>();
            loader.Setup(l => l.Load("dexter")).Returns(new User { Name = "dexter", Password = "killer" });

            string token = new UserAccessChecker(loader.Object, new Tools.RandomString()).Check(userToCheck);

            Assert.Equal(32, token.Length);
        }

        [Fact]
        public void Test_known_Name_With_Wrong_Password_should_return_Error_Access_Denied()
        {
            IUser userToCheck = new User { Name = "dexter", Password = "morgan" };
            Mock<IUserByNameLoader> loader = new Mock<IUserByNameLoader>();
            loader.Setup(l => l.Load("dexter")).Returns(new User { Name = "dexter", Password = "killer" });

            Assert.Throws<AccessDeniedException>(() => new UserAccessChecker(loader.Object, new Tools.RandomString()).Check(userToCheck));
        }
    }
}

File: Go/STED/Usecases/UserAccessChecker_test.go
package Usecases

import (
	"fmt"
	"serenityapi/Domain"
	"testing"

	"github.com/stretchr/testify/assert"
)

var mockLoader IUserByNameLoader = new(Mock_UserByNameLoader)

func Test_No_Name_No_Password_should_return_Error_Access_Denied(t *testing.T) {
	user := Domain.User{Name: "", Password: ""}

	_, err := UserAccessChecker(user, mockLoader)

	assert.EqualError(t, err, ACCESS_DENIED_ERROR)
}

func Test_Unknown_Name_should_return_Error_Access_Denied(t *testing.T) {
	user := Domain.User{Name: "debra", Password: ""}

	_, err := UserAccessChecker(user, mockLoader)

	assert.EqualError(t, err, ACCESS_DENIED_ERROR)
}

func Test_known_Name_With_Known_Password_should_return_Valid_Token(t *testing.T) {
	user := Domain.User{Name: "dexter", Password: "killer"}

	token, err := UserAccessChecker(user, mockLoader)

	assert.NoError(t, err)
	assert.Equal(t, 32, len(token), fmt.Sprintf("Token length should be 32 be was %d", len(token)))
}

func Test_known_Name_With_Wrong_Password_should_return_Error_Access_Denied(t *testing.T) {
	user := Domain.User{Name: "dexter", Password: "morgan"}

	_, err := UserAccessChecker(user, mockLoader)

	assert.EqualError(t, err, ACCESS_DENIED_ERROR)
}
